### PR TITLE
Add metrics to replicator

### PR DESCRIPTION
**Pull request by nickva**
Added metrics for clustering, doc processing and job scheduler.

There are 3 guages for averge times for job scheduler:
- avg_running : average time of all the current running processes
- avg_pending : average time waiting to run of all current stopped
  processed (these don't include crashed)
- avg_crashed : average time crashed jobs are being forced
  to wait

### CODE REVIEW

File: src/couch_replicator_clustering.erl [hunk 1]
```diff
@@ -120,6 +120,7 @@ handle_info({nodeup, Node}, State) ->
 handle_info({nodedown, Node}, State) ->
     Timer = new_timer(interval(State)),
     couch_replicator_notifier:notify({cluster, unstable}),
+    couch_stats:update_gauge([couch_replicator, cluster_is_stable], 0),
```
Comment (sagelywizard) [inline src/couch_replicator_clustering.erl:125]: Perhaps add another `update_gauge` to `0` in `init`?

**Comment by sagelywizard** [comment]
Maybe we should add something to indicate how long jobs are running for and how long jobs are being stopped for? Also, how long crashed jobs are being suspended for. Latency histograms?

**Comment by sagelywizard** [comment]
The max gauges seem a little superfluous to me. Thoughts about keeping/removing them @rnewson?

File: src/couch_replicator_clustering.erl (hunk 1 above)
Comment (nickva) [inline src/couch_replicator_clustering.erl:125]: Yap, good point. Added it. Also noticed forgot one on `nodeup` a well. So added that too.

**Comment by nickva** [comment]
@sagelywizard I added max_\* gauges to super-impose on and compare with actual running/churn,  Mainly to verify the correctness of our scheduler.  But can remove them, they are probably superfluous.

**Comment by nickva** [comment]
I thought of adding histograms initially but there were a few issues:

 1) Now sure which ones to add. It seems for running jobs, add histogram of time running, for those that are stopped, time stopped, for those that are crashed -- time since last crash?

 2) The default histograms we get with `couch_stats` is hard coded to be of type  `slide_uniform` with a time window of 10 seconds, and 1024 samples. If I understand how it works, it will drop off entries older than 10 seconds, that is less than the reschedule interval. 

So can probably do our own average calculation and turn it into a gauge, or use folsom directly, by-passing couch_stats.

File: src/couch_replicator_scheduler.erl [hunk 2]
```diff
@@ -473,6 +500,56 @@ update_history(Job, Type, When, State) ->
     Job#job{history = History1}.
 
 
+-spec update_running_jobs_stats() -> ok.
+update_running_jobs_stats() ->
+    Acc0 = #stats_acc{now = os:timestamp()},
+    AccR = ets:foldl(fun stats_fold/2, Acc0, ?MODULE),
+    #stats_acc{
+        pending_t = PendingSumUSec, pending_n = PendingN,
```
Comment (sagelywizard) [inline src/couch_replicator_scheduler.erl]: It's my personal preference that each of these has its own line. So take that for what you will. :D

Comment (nickva) [inline src/couch_replicator_scheduler.erl]: Yap, agree

File: src/couch_replicator_scheduler.erl [hunk 3]
```diff
@@ -473,6 +500,56 @@ update_history(Job, Type, When, State) ->
...
+    couch_stats:update_gauge([couch_replicator, jobs, crashed], CrashedN),
+    couch_stats:update_gauge([couch_replicator, jobs, avg_pending], PendingAvg),
+    couch_stats:update_gauge([couch_replicator, jobs, avg_running], RunningAvg),
+    couch_stats:update_gauge([couch_replicator, jobs, avg_crashed], CrashedAvg),
+    ok.
+
+
+-spec stats_fold(#job{}, #stats_acc{}) -> #stats_acc{}.
+stats_fold(#job{pid = undefined, history = []}, Acc) ->
+    Acc;
```
Comment (sagelywizard) [inline src/couch_replicator_scheduler.erl]: Shouldn't this count as pending?

**Comment by sagelywizard** [comment]
Hrm. For the stop/start/crashed, how about we just record that in a histogram when there's a state transition. Like, whenever a job stops, `couch_stats:update_histogram([couch_replicator, job_time, stopped], timer:now_diff(A, B)),`?

File: src/couch_replicator_scheduler.erl [hunk 4]
```diff
@@ -298,7 +312,11 @@ backoff_micros(CrashCount) ->
 
 -spec add_job_int(#job{}) -> boolean().
 add_job_int(#job{} = Job) ->
-    ets:insert_new(?MODULE, Job).
+    InsertRes = ets:insert_new(?MODULE, Job),
+    TotalJobs = ets:info(?MODULE, size),
+    couch_stats:update_gauge([couch_replicator, jobs, total], TotalJobs),
```
Comment (rnewson) [inline src/couch_replicator_scheduler.erl]: if insert_new returned false then nothing changed, is it useful to update metrics in that case?

File: src/couch_replicator_scheduler.erl [hunk 5]
```diff
@@ -298,7 +312,11 @@ backoff_micros(CrashCount) ->
...
+    couch_stats:increment_counter([couch_replicator, jobs, adds]),
```
Comment (rnewson) [inline src/couch_replicator_scheduler.erl]: this is tracking the number of calls to add_job_int or the number of times we add a job? If the latter, it is wrong if InsertRes is false.

File: src/couch_replicator_scheduler.erl (hunk 4 above)
Comment (nickva) [inline src/couch_replicator_scheduler.erl]: Yeah that was to track number of times other code tried to add jobs

File: src/couch_replicator_scheduler.erl (hunk 3 above)
Comment (nickva) [inline src/couch_replicator_scheduler.erl]: Agree but we don't have any history. Not sure what to do about it. So just skipped it.

**Comment by pszczypka** [comment]
Just to play devils advocate - are we absolutely sure all the proposed metrics are actually useful?

**Comment by sagelywizard** [comment]
All the ones which didn't seem useful to me were removed.

**Comment by pszczypka** [comment]
Jobs total would appear to be a function of the other metrics.

**Comment by nickva** [comment]
@pszczypka Hey Paul :-)  Not sure yet. We can remove them if they are not useful. Once we play with an actual cluster and see what they look, it is hard to say yet. Some metrics like total is there to verify if our code works correctly. For example if pending + crashed + running < total might indicate a problem. Or would indicate there is a duplicate replication running on another node.

**Comment by pszczypka** [comment]
Fair enough assuming this is for testing. I'm urging caution and suggesting that the only the minimum required metrics are deployed. All too often a test becomes production becomes standard and no one ever uses it.

At least you're not abusing the histograms.

**Comment by sagelywizard** [comment]
Currently pending + crashed + running is less than total, because unstarted jobs jobs are not counted as pending. It's still in flux.

File: src/couch_replicator_scheduler.erl [hunk 6]
```diff
@@ -337,7 +355,11 @@ stop_job_int(#job{} = Job, State) ->
 
 -spec remove_job_int(#job{}) -> true.
 remove_job_int(#job{} = Job) ->
-    ets:delete(?MODULE, Job#job.id).
+    DeleteRes = ets:delete(?MODULE, Job#job.id),
+    TotalJobs = ets:info(?MODULE, size),
+    couch_stats:update_gauge([couch_replicator, jobs, total], TotalJobs),
+    couch_stats:increment_counter([couch_replicator, jobs, removes]),
```
Comment (rnewson) [inline src/couch_replicator_scheduler.erl]: we increment remove even if we didn't remove anything. same solution as for add_job, move this to the {ok, Job} clause of add_job

**Comment by rnewson** [comment]
+1 after fixing the remove bug.

**Pull request merged**